src/algebra/field.lean

```diff
@@ -256,3 +256,19 @@ protected lemma injective : function.injective f := f.injective_iff.2 $ λ x, f.
 end
 
 end ring_hom
+
+section noncomputable_defs
+
+variables {R : Type*} [nontrivial R]
+
+/-- Constructs a `division_ring` structure on a `ring` consisting only of units and 0. -/
+noncomputable def division_ring_of_is_unit_or_eq_zero [ring R]
+  (h : ∀ (a : R), is_unit a ∨ a = 0) : division_ring R :=
+{ .. (group_with_zero_of_is_unit_or_eq_zero h), .. (infer_instance : ring R) }
```

Comment:
I recall one definition projecting fields from `infer_instance` that caused the unifier to struggle, so including the fields directly like this should be safer.
```suggestion
noncomputable def division_ring_of_is_unit_or_eq_zero [hR : ring R]
  (h : ∀ (a : R), is_unit a ∨ a = 0) : division_ring R :=
{ .. (group_with_zero_of_is_unit_or_eq_zero h), .. hR }
```